### ISSUE TITLE
ConditionalTerm

### ISSUE THREAD

**Issue by silviabonfanti** [uasm/language-definiton#10]
I would like to propose an extension for the ConditionalTerm.

Is it possible to add 'endif'? (as optional) like in ConditionalRule?

Old ConditionalTerm
ConditionalTerm ::= Term '?' Term ':' Term | 'if' Term 'then' Term 'else' Term

New ConditionalTerm
ConditionalTerm ::= Term '?' Term ':' Term | 'if' Term 'then' Term 'else' Term 'endif'?

This is usefull to us for the autoformatting function.

**Comment by mstegmaier** [comment]
I agree and it might desirable to even make it mandatory to avoid ambiguity. For example see:
`if b then 0 else 1 + 1`
what will this return in case b is true?
Will it return 0 because it will be interpreted as
`if b then 0 else (1 + 1)`
or as
`(if b then 0 else 1) + 1`
matters can get even worse when nesting ConditionalTerms
`if b then 0 else if c then 1 else 2 + 1`
Is this to be interpreted as
`if b then 0 else if c then 1 else (2 + 1)`
or as
`if b then 0 else ((if c then 1 else 2) + 1)`
or as
`(if b then 0 else if c then 1 else 2) + 1`

**Comment by silviabonfanti** [comment]
I agree with your examples. I think that it's better to make it mandatory.

We could have the same problem also with the ConditionalRule. 
At the moment 'endif' is optional in ConditionalRule.
Do you think that we should make it mandatory?

**Comment by mstegmaier** [comment]
Actually it's not quite the same problem.
The problem with ConditionalRule only occurs because the 'else' is optional which can cause a so-called [dangling-else](https://en.wikipedia.org/wiki/Dangling_else).
The problem with ConditionalTerm is a problem of operator precedences. It depends on the precedence of this ternary operator compared to the binary operators.
Usually binary operators have lower precedence than unary operators so we could argue that ternary operators should have even less precedence. But there is already an exception to this rule, i.e., the power operator '^' which is a binary operator with a precedence higher than the unary operators.
By enforcing the 'endif' at the end of the term we kinda turn this ternary operator into something that lets us avoid the question of precedences in this case.

If we make the 'endif' of the ConditionalRule mandatory we will need to think about making all 'end*' mandatory.

**Comment by garganti** [comment]
For conditional rule, it is true that the endif is not necessary because we can agree that the dangling else (https://en.wikipedia.org/wiki/Dangling_else) refers to the last nested if. However it would be useful if one wants to make an else for a previous inner rule (in terms you can use parenthesis, but in rules not).
For instance, if I write
`if a then if b then R1 else R2`
according to our semantics of the dangling else, R2 refers to if b. If I want to make an else for the first "if a", how can I do that? Something like:
`if a then par if b then R1 endpar else R2`? 
Isn't too complex and wrong since par must have at least two rules? 
Adding an "endif" would solve the problem and solve also the dangling else problem.
I could write:
`if a then if b then R1 else R2 endif endif` or
`if a then if b then R1 endif else R2 endif`

**Comment by mstegmaier** [comment]
Your comment is a bit unrelated. This issue discusses the need for 'endif' for the ConditionalTerm not the ConditionalRule. While everyone would agree on at least having an optional 'endif' it's still to be decided whether it should even be mandatory for the ConditionalTerm. During the discussion, our attention was drawn to the optional 'endif' of the ConditionalRule and we wondered whether it should be made mandatory as well in case we made the 'endif' of the ConditionalTerm mandatory. But if we did that we would need to think about making the 'end*' of all rules that have one mandatory.

Your comment seems to discuss whether the 'endif' of the ConditionalRule should be omitted or not. But that was never part of our discussion as it would be kind of reckless to do so (as you have noticed).

**Comment by garganti** [comment]
Yes, my comment is clearly about the ConditionalRule - not sure it is unrelated. Anyway, we have opened a new issue for conditional rule.
